### ISSUE TITLE
🐞 Padim - Difference in Engine prediction vs ONNX or Torch prediction

### ISSUE THREAD

**Issue by mbignotti** [open-edge-platform/anomalib#2944]
### Describe the bug

Hi,
I am sorry to bother you, but I can't get an exported ONNX model to work at inference time as inside the `engine.predict()` method. In particular, when I use the `engine.fit(model, datamodule)` method, followed by `engine.predict(model, datamodule)` I get a reasonable result, that seems correct.
When I try to use the exported ONNX model, or even the same torch based model without using the engine and the datamodule (a single forward pass), I get a different result that seems just wrong.

I have searched into the code base, but I did not manage to understand what is it that the Engine is doing differently.

I have attached a reproducible example based on the MVTechAD dataset. In the `experiment.ipynb` notebook you can see that the ONNX and the pytorch model return the same result. The inference on the same image with the engine, however, yields a different result.

Note: In order to re-run the notebook it is necessary to re-download the dataset and to change the paths to the images inside the code.

### Dataset

MVTecAD

### Model

PADiM

### Steps to reproduce the behavior

1. Unzip the file
2. Run `uv sync` and `uv lock` to install the virtual env
3. Check the `experiment.ipynb` notebook

[sample.zip](https://github.com/user-attachments/files/22214569/sample.zip)

### OS information

OS information:
- OS: Ubuntu 24.04
- Python version: 3.12.0
- Anomalib version: 2.1.0
- PyTorch version: 2.8.0
- CUDA/cuDNN version: 12.8
- GPU models and configuration: GeForce RTX 3070

### Expected behavior

ONNX and Torch models should yield the same output as `engine.predict()`, provided the input image is pre-processed correctly.

### Screenshots

<img width="788" height="348" alt="Image" src="https://github.com/user-attachments/assets/af380e17-a9c0-4849-8120-b59d02c61e0b" />

<img width="1145" height="805" alt="Image" src="https://github.com/user-attachments/assets/91278be6-4d5d-421c-b5ea-8f2a91e82636" />

### Pip/GitHub

pip

### What version/branch did you use?

2.1.0

### Configuration YAML

```yaml
None
```

### Logs

```shell
None
```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by DeepKnowledge1** [comment]
@mbignotti 


I faced the same problem before. The mismatch doesn’t actually come from the ONNX export — ONNX and PyTorch give the same raw results. The difference is that engine.predict() in anomalib adds extra preprocessing and postprocessing steps (normalization, resizing, Gaussian blur smoothing, etc.). That’s why the output from the engine looks different from running a single forward or the exported model directly.

If you want a cleaner setup, you can check out my implementation here: [AnomaVision](https://github.com/DeepKnowledge1/AnomaVision). I exported models into different formats (Torch, ONNX, etc.) and they all give consistent results across inference, without hidden differences. In my benchmarks, it actually outperformed anomalib.

Here’s a demo video: https://youtu.be/6x19w0XtNPk.

**Comment by mbignotti** [comment]
Hi @DeepKnowledge1,
Thank you for your comment!
However, I think that this should have been solved with the Anomalib 2.0 release. As described in the [docs](https://anomalib.readthedocs.io/en/v2.0.0/markdown/guides/how_to/models/pre_processor.html), both pre and post processing should now be included in the onnx graph.
I also confirmed this while debugging the pytorch model. Hence, I don't think that's the issue, unless there are other callbacks called by the engine that are doing some more pre-processing.
I would rather wait for @ashwinvaidya17 or some other maintainer to answer on this.

Thanks again!

**Comment by samet-akcay** [comment]
Thought this has been addressed with https://github.com/open-edge-platform/anomalib/pull/2843. We'll double check whether this issue still persists

**Comment by mbignotti** [comment]
Hi!
I was wondering if you got the chance to take a look at the issue.
I tested with version 2.2.0 by using the same code attached, and the problem seem to still persist.

Thanks again!

**Comment by tailyer** [comment]
@mbignotti : thanks for your detailed bug report! 
@samet-akcay I got the same problem when I tried to use torch inferencer (Fastflow model - Anomalib v2.2.0). Seems like the bug is still there!

**Comment by InvalidSyntaxxx** [comment]
same problem （Dinomaly model - Anomalib v2.2.0 ）when export to OpenVINO model

**Comment by MHzl** [comment]
Same problem. Fastflow with 2.2.0. :(

**Comment by BlankHead2004** [comment]
@mbignotti can you share your trained onnx model?

**Comment by mbignotti** [comment]
@BlankHead2004, sorry. I am having an issue in downloading the MVTecAD dataset now. Not sure if it's the link or if I have some kind of firewall rule that is blocking the access to the dataset (I do not have my personal laptop with me at the moment). Can you try run the notebook attached?

**Comment by rajeshgangireddy** [comment]
> [@BlankHead2004](https://github.com/BlankHead2004), sorry. I am having an issue in downloading the MVTecAD dataset now. Not sure if it's the link or if I have some kind of firewall rule that is blocking the access to the dataset (I do not have my personal laptop with me at the moment). Can you try run the notebook attached?

Hi @mbignotti 
About the mvtecad dataset 404 error, see #3244

**Comment by mbignotti** [comment]
Please find attached the trained model.
[model.zip](https://github.com/user-attachments/files/24713488/model.zip)

**Comment by terrarier2111** [comment]
@ashwinvaidya17 any updates on this?

**Comment by jgespino** [comment]
This issue has been inactive for a while. I wanted to check in and see if this is still relevant or if it can be closed. If the issue has been resolved or is no longer relevant, we can go ahead and close it. We'll close this in 7 days if there's no response.

**Comment by mbignotti** [comment]
Hi @jgespino,
This issue is still relevant. As you can see from the comments above, many users can't deploy models to production and others have opened similar issues. There is ongoing discussion in #3261 as well. Please keep it open.

**Comment by samet-akcay** [comment]
FYI @ashwinvaidya17

**Comment by ashwinvaidya17** [comment]
Is this still the case on the main branch? I'll investigate.

**Comment by mbignotti** [comment]
Hi @ashwinvaidya17!
Did you get the chance to see if the issue still persists in the main branch? I tried with version 2.2.0 and with different versions of onnx/pytorch, but the output is still wrong. Actually, I don't think it's a dependency issue, otherwise even the `Engine.predict/test` methods would yield wrong results, but they are correct! It must be something that the engine is doing that is not being replicated with the bare torch/onnx models.

Thanks!

**Comment by ashwinvaidya17** [comment]
I was not able to reproduce it locally on the main branch. I guess once we make the release it should fix it.

**Comment by mbignotti** [comment]
Actually, I have just run the notebook attached at the beginning of this issue with the unreleased version of anomalib, and I'm still seeing major differences between engine/onnx/torch. Could you share the code you are using?

**Comment by ashwinvaidya17** [comment]
Ah didn't see the attached .zip. I was able to isolate the issue. The issue becomes more pronounced when the resizing is more extreme. The problem is due to anti aliasing.
https://github.com/open-edge-platform/anomalib/blob/8b09016c83eb948389e1ed2aee020f990a9485e8/src/anomalib/pre_processing/utils/transform.py#L51
https://github.com/open-edge-platform/anomalib/blob/8b09016c83eb948389e1ed2aee020f990a9485e8/src/anomalib/pre_processing/pre_processor.py#L79

`engine.predict` uses transform which has `anti_aliasing=True`

whereas when exporting `forward(...)` is called which sets `anti_aliasing=False`
https://github.com/open-edge-platform/anomalib/blob/8b09016c83eb948389e1ed2aee020f990a9485e8/src/anomalib/pre_processing/pre_processor.py#L144
This is because `onnx` export does not support `anti_aliasing`.

one approach to get the same numbers would be

```diff
class Preprocessor(...):
    ...
    def on_predict_batch_start(...
        ...
-      if self.transform:
-           batch.image, batch.gt_mask = self.transform(batch.image, batch.gt_mask)
+      if self.export_transform:
+           batch.image, batch.gt_mask = self.export_transform(batch.image, batch.gt_mask)
```

Though personally I feel if `engine.predict` supports anti-aliasing then we should keep it and hope in the future that onnx supports it as well.

Let me know if this solves your issue

**Comment by mbignotti** [comment]
I see onnx `Resize` operator has an `antialias` argument since version 18: https://onnx.ai/onnx/operators/onnx__Resize.html

However I am not super familiar with the conversion from torchvision transforms to onnx operators, and I am not sure they are the same arguments.

**Comment by mbignotti** [comment]
Anyways, I confirm that by setting `antialias=False` in `v2.Resize`, the engine and the onnx/torch outputs match exactly.

**Comment by ashwinvaidya17** [comment]
Here is my reproducer. Can you try setting `antialias=True/False`

```python
import torch
from torch import nn
from torchvision.transforms import Compose, Resize


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.transform = Compose([Resize(256, antialias=True)])

    def forward(self, x):
        x = self.transform(x)
        return x


model = Model()
dummy_input = torch.randn(1, 3, 1, 1)
dynamic_axes = {"input": {0: "batch_size"}, "output": {0: "batch_size"}}
torch.onnx.export(
    model,
    dummy_input,
    "model.onnx",
    input_names=["input"],
    output_names=["output"],
    dynamo=False,
    dynamic_axes=dynamic_axes,
)
```

Export fails when dynamic axes is set

**Comment by mbignotti** [comment]
Yes, `antialias=True` seems to be unsupported even with `opset_version=19`. Then I guess the only solution for now is to manually set `antialias=False` in the `PreProcessor`.

Thanks for finding the issue!
